dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/TasksRunner.java

```diff
@@ -0,0 +1,178 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.apache.commons.lang3.time.DurationFormatUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** @author ishmum */
+public class TasksRunner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TasksRunner.class);
+  private static final Logger PROGRESS_LOG = LoggerFactory.getLogger("progress-logger");
+
+  private AtomicInteger atomicNumberOfCompletedTasks;
+  private final int totalNumberOfTasks;
+  private final Stopwatch stopwatch;
+
+  private final TaskRunContext context;
+  private final TaskSetState.Impl state;
+  private final List<Task<?>> tasks;
+
+  public TasksRunner(
+      OutputHandleFactory sinkFactory,
+      Handle handle,
+      int threadPoolSize,
+      @Nonnull TaskSetState.Impl state,
+      List<Task<?>> tasks) {
+    context = createContext(sinkFactory, handle, threadPoolSize, state);
+    this.state = state;
+    this.tasks = tasks;
+    totalNumberOfTasks = countTasks(tasks);
+    stopwatch = Stopwatch.createStarted();
+    atomicNumberOfCompletedTasks = new AtomicInteger();
+  }
+
+  private TaskRunContext createContext(
+      OutputHandleFactory sinkFactory, Handle handle, int threadPoolSize, Impl state) {
+    return new TaskRunContext(sinkFactory, handle, threadPoolSize) {
+      @Override
+      public TaskState getTaskState(Task<?> task) {
+        return state.getTaskState(task);
+      }
+
+      @Override
+      public <T> T runChildTask(Task<T> task) throws MetadataDumperUsageException {
+        return handleTask(task);
+      }
+    };
+  }
+
+  public void run() throws MetadataDumperUsageException {
+    for (Task<?> task : tasks) {
+      handleTask(task);
+    }
+  }
+
+  @CheckForNull
+  private <T> T handleTask(Task<T> task) throws MetadataDumperUsageException {
+    T t = runTask(task);
+    if (!(task instanceof TaskGroup)) {
+      atomicNumberOfCompletedTasks.getAndIncrement();
+    }
+    logProgress();
+    return t;
+  }
+
+  private void logProgress() {
+    int numberOfCompletedTasks = atomicNumberOfCompletedTasks.get();
+
+    long averageTimePerTaskInMillisecond =
+        stopwatch.elapsed(TimeUnit.MILLISECONDS) / numberOfCompletedTasks;
+
+    int percentFinished = numberOfCompletedTasks * 100 / totalNumberOfTasks;
+    String progressMessage = percentFinished + "% Completed";
+
+    int remainingTasks = totalNumberOfTasks - numberOfCompletedTasks;
+    long remainingTimeInMillis = averageTimePerTaskInMillisecond * remainingTasks;
+
+    if (numberOfCompletedTasks > 10) {
+      String estimationMessage =
+          "Approximately "
+              + DurationFormatUtils.formatDurationWords(remainingTimeInMillis, true, true)
+              + " remaining";
+      PROGRESS_LOG.info(progressMessage + "; " + estimationMessage);
```

Comment:
I would suggest changing the semicolon `"; "` to a dot `". "`.

Also, consider changing the code by adding the line:
```
progressMessage += ". " + estimationMessage
```

to log only once and eliminate the `else` block below.